### ISSUE TITLE
TypeScript fork

### ISSUE THREAD

**Issue by jonnyelliot** [joshjung/trie-search#30]
I thought I should let you guys know that I have been working a TypeScript version of this library. You may see a version of it [here](https://github.com/commitd/trie-search).

While I have forked the project for my own use case, I was wondering if there would be interest in merging this upstream?

Cheers to @joshjung and thanks for the great library 🙂

**Comment by chetbox** [comment]
Thanks for this @jonnyelliot! This is a great library and a Typescript version is exactly what we were looking for.

@jonnyelliot and @joshjung I'd be very happy the Typescript version could be merged here.

**Comment by joshjung** [comment]
Hey, I should have time this weekend to take care of this! Thanks for getting this done :D

**Comment by joshjung** [comment]
@jonnyelliot Well I finally got around to this, and realized you rewrote it, which is amazing. I thought it was originally just the definitions, but the code is now TypeScript. I'm not really sure how to proceed as there are so many people using this now that if I just switch to TypeScript it might break a lot of things for people. Hmmm... how confident are you that the TypeScript version will be backwards compatible with the old import methodology?

Basically I'm wondering if we should include the original raw JS and the TypeScript at the same time and just include a way to use either one. Unfortunately I haven't done that before. Thoughts?

**Comment by chetbox** [comment]
It might be worth a major version number change if the API has changed at all?

Libraries built with Typescript ship vanilla JS and also `.d.ts` type definition files by convention. If the API is the same as the JS version users will be able to carry on using future versions with no noticeable changes.

**Comment by adalinesimonian** [comment]
Taking a closer look at the typescript fork, it looks like it makes a few type constraints that don't match the behaviour of the non-TS library.

----

Specifically, the accumulator in the reducer function passed to `get`/`search` is constrained to an array of objects of the same type as the type of values in the trie. However, in the JS library, you can use any type of accumulator, even if its type differs.

```js
const trie = new TrieSearch()

trie.map('some string', 'value')
// ...

trie.get('string', (accumulator, phrase, matches, trie) =>
  matches.reduce(
    (lines, match) => `${lines}${trie.getId(match)}\n`,
    accumulator || ''
  )
)
```

----

Another minor difference, one that's more debatable, is that the type of `item` expected by `getId(item)` is also constrained to the type of values in the trie. When using trie-search, you can technically create a trie with an `idFieldOrFunction` that expects a different type of item since only reducers call `getId`, and reducers can be user-written. `UNION_REDUCER` is just one implementation, one that expects any implementation of a function for `idFieldOrFunction` to accept the same type as the values in the trie for `item`.

Here's one example where, to get an ID, the item passed to `getId` has been processed through another function that returns a different type, still corresponding to the same item.

```js
const trie = new TrieSearch(undefined, {
  // item is expected to be a DatabaseEntry instance
  idFieldOrFunction: item => item && item.getKey()
})

trie.map('some string', 'key 1')

trie.get('string', (accumulator, phrase, matches, trie) =>
  matches.reduce((acc, match) => {
    // item passed to getItem is a DatabaseEntry instance
    acc.push(trie.getId(db.getItem(match)))
    return acc
  }, accumulator || [])
)
```

`getItem`'s implementation is technically agnostic, when `idFieldOrFunction` is a function, with respect to the type of the item that is passed to it:

```js
return typeof this.options.idFieldOrFunction === 'function'
  ? this.options.idFieldOrFunction(item)
  : item[this.options.idFieldOrFunction];
```

Now whether or not this is good practice is a completely different issue. Personally, it feels wrong to put functionality outside of the scope of the type of value in the trie into that very same trie. So, it might indeed be a good idea to constrain the type of `item` such that it must be of the same type as what's in the trie.

----

For reference, here are the type definitions that I wrote for trie-search that I'm using in a project I'm working on. In making these definitions I was looking only to write constraints correlated to existing constraints in the library, without adding any additional constraints.

```typescript
/* eslint-disable no-underscore-dangle */
declare module 'trie-search' {
  type KeyFields = string | string[] | KeyFields[]

  type TrieSearchOptions<T> = {
    ignoreCase?: boolean
    maxCacheSize?: number
    cache?: boolean
    splitOnRegEx?: RegExp
    splitOnGetRegEx?: RegExp
    min?: number
    keepAll?: boolean
    keepAllKey?: string
    idFieldOrFunction?: string | ((item: T) => string)
    expandRegexes?: { regex: RegExp; alternate: string }[]
    insertFullUnsplitKey?: boolean
  }

  type Reducer<A, T, I = T> = (
    accumulator: A | undefined,
    phrase: string,
    matches: T[],
    trieSearch: TrieSearch<T, I>
  ) => A | undefined

  export default class TrieSearch<T, I = T> {
    constructor(keyFields?: KeyFields, options?: TrieSearchOptions<I>)

    add(obj: T, customKeys?: KeyFields | number): void

    expandString(value: string): string[]

    addAll(arr: T[], customKeys?: KeyFields | number): void

    reset(): void

    clearCache(): void

    cleanCache(): void

    addFromObject(obj: T, valueField: string): void

    map(key: string, value: T): void

    keyToArr(key: string): string[]

    findNode(key: string): { [key: string]: T[] }

    _getCacheKey(phrase: string, limit?: number): string

    _get(phrase: string, limit?: number): T[]

    get(
      phrases: string | string[],
      reducer?: null | undefined,
      limit?: number
    ): T[]

    get<A>(
      phrases: string | string[],
      reducer?: Reducer<A, T, I>,
      limit?: number
    ): A | undefined

    search(phrases: string | string[], reducer: null | undefined): T[]

    search<A>(
      phrases: string | string[],
      reducer: Reducer<A, T, I>
    ): A | undefined

    getId(item: I): string

    static UNION_REDUCER: <V>(
      accumulator: V[] | undefined,
      phrase: string,
      matches: V[],
      trieSearch: TrieSearch<V, V>
    ) => V[]
  }
}
```

----

The constraints on the `UNION_REDUCER` function ensure that it can only be used on tries where the values in the trie are the same type as the `item` expected by any possible id function. For example:

```typescript
const trie = new TrieSearch<MyItem, string>(undefined, {
  idFieldOrFunction: (item: string) => item
})

trie.map('some string', someItem)
// ...

trie.get<MyItem[]>('string', TrieSearch.UNION_REDUCER)
// compilation error: Argument of type '<V>(accumulator: V[] | undefined,
// phrase: string, matches: V[], trieSearch: TrieSearch<V, V>) => V[]' is not
// assignable to parameter of type 'Reducer<MyItem[], MyItem, string>'.
```

as compared to:

```typescript
const trie = new TrieSearch<MyItem>(undefined, {
  idFieldOrFunction: (item: MyItem) => item.getKey()
})

trie.map('some string', someItem)
// ...

trie.get<MyItem[]>('string', TrieSearch.UNION_REDUCER)
// no errors
```

However, if the behaviour of allowing `item` to be a different type is a bad pattern (I certainly think it is) then a constraint to prevent this bad pattern entirely avoids the necessity of this check.

**Comment by joshjung** [comment]
It is clear to me I need to take a closer and more serious look at this TypeScript impl. I will confess I am just _super nervous_ about making any changes to the release because it is has been working so well with so few bugs for so long and I'm afraid if I touch it I'm going to have an onslaught of bug reports.

Until I get time to take a much longer and harder look at this, I highly recommend those who want Typescript, please use the fork provided by @jonnyelliot .

My plan right now is:

* Take a super close line-by-line look at the TypeScript rewrite 
* Consider removing the dependency on `hasharray`
* Merge in the fork, with a major version bump
* Re-release and update the Readme with instructions to use the former major version if there are breaking changes

I am moving back to the US here this next week so I've been pretty busy but will hopefully have time in the next month to complete this.

**Comment by adalinesimonian** [comment]
@jonnyelliot A much less intrusive fix may be just to provide type declarations in the package. That would not require any changes to implementation (or at least very minor ones) and would allow Typescript users to use the package without any issue. That's what we're doing over at Stylelint — our next major version will provide types in the package, and we didn't have to rewrite the project to achieve this. If you'd like to rewrite the package in Typescript, I feel it would be a better candidate for a major version than a minor.

**Comment by lnvglr** [comment]
Hey @joshjung are there any news on your plans to include the type definitions?

**Comment by joshjung** [comment]
I'm going to get this done this week, at least that is the plan. Startup life had me strapped, but I've got a ton of free time now :D

**Comment by joshjung** [comment]
@jonnyelliot Would love to touch base to get your fork back into the main repo here. You still around working on it?

**Comment by joshjung** [comment]
Since I did not hear back from @jonnyelliot I went ahead and followed the suggestion of @adalinesimonian and simply added a TypeScript definition file and switch to using ts-jest for running my tests and all seems to be working beautifully.

Apologies to anyone who cared that this took so long, I kept forgetting with startup life and all that jazz :D
